File: camerax_android/lib/src/core/focus_metering_action_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class FocusMeteringAction$BuilderImpl
    implements FocusMeteringAction$Builder {
  final FocusMeteringActionBuilderProxyApi api;

  FocusMeteringAction$BuilderImpl.internal(this.api);

  @override
  Future<FocusMeteringAction$Builder> addPoint(
    MeteringPoint point, [
    List<FocusMeteringAction$MeteringMode>? meteringModes,
  ]) => meteringModes == null
      ? api.addPoint1(point.api).then((e) => e.impl)
      : api
            .addPoint2(point.api, meteringModes.map((e) => e.api).toList())
            .then((e) => e.impl);

  @override
  Future<FocusMeteringAction$Builder> disableAutoCancel() =>
      api.disableAutoCancel().then((e) => e.impl);

  @override
  Future<FocusMeteringAction$Builder> setAutoCancelDuration(
    Duration duration,
  ) => api
      .setAutoCancelDuration(duration.inMicroseconds, TimeUnitApi.microseconds)
      .then((e) => e.impl);

  @override
  Future<FocusMeteringAction> build() => api.build().then((e) => e.impl);
}

final class FocusMeteringActionImpl implements FocusMeteringAction {
  final FocusMeteringActionProxyApi api;

  FocusMeteringActionImpl.internal(this.api);

  @override
  Future<Duration> getAutoCancelDuration() => api
      .getAutoCancelDurationInMillis()
      .then((e) => Duration(milliseconds: e));

  @override
  Future<List<MeteringPoint>> getMeteringPointsAe() =>
      api.getMeteringPointsAe().then((e) => e.map((e1) => e1.impl).toList());

  @override
  Future<List<MeteringPoint>> getMeteringPointsAf() =>
      api.getMeteringPointsAf().then((e) => e.map((e1) => e1.impl).toList());

  @override
  Future<List<MeteringPoint>> getMeteringPointsAwb() =>
      api.getMeteringPointsAwb().then((e) => e.map((e1) => e1.impl).toList());

  @override
  Future<bool> isAutoCancelEnabled() => api.isAutoCancelEnabled();
}

final class FocusMeteringActionChannelImpl extends FocusMeteringActionChannel {
  @override
  FocusMeteringAction$Builder createBuilder(
    MeteringPoint point, [
    List<FocusMeteringAction$MeteringMode>? meteringModes,
  ]) {
    final api = meteringModes == null
        ? FocusMeteringActionBuilderProxyApi.new1(point: point.api)
        : FocusMeteringActionBuilderProxyApi.new2(
            point: point.api,
            meteringModes: meteringModes.map((e) => e.api).toList(),
          );
    return FocusMeteringAction$BuilderImpl.internal(api);
  }
}

extension FocusMeteringAction$MeteringModeX
    on FocusMeteringAction$MeteringMode {
  FocusMeteringActionMeteringModeApi get api =>
      FocusMeteringActionMeteringModeApi.values[index];
}

extension FocusMeteringActionBuilderProxyApiX
    on FocusMeteringActionBuilderProxyApi {
  FocusMeteringAction$Builder get impl =>
      FocusMeteringAction$BuilderImpl.internal(this);
}

extension FocusMeteringActionX on FocusMeteringAction {
  FocusMeteringActionProxyApi get api {
    final impl = this;
    if (impl is! FocusMeteringActionImpl) throw TypeError();
    return impl.api;
  }
}

extension FocusMeteringActionProxyApiX on FocusMeteringActionProxyApi {
  FocusMeteringAction get impl => FocusMeteringActionImpl.internal(this);
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/FocusMeteringActionImpl.kt
package dev.zeekr.camerax_android.core

import androidx.camera.core.FocusMeteringAction
import androidx.camera.core.MeteringPoint
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.FocusMeteringActionMeteringModeApi
import dev.zeekr.camerax_android.PigeonApiFocusMeteringActionBuilderProxyApi
import dev.zeekr.camerax_android.PigeonApiFocusMeteringActionProxyApi
import dev.zeekr.camerax_android.TimeUnitApi
import dev.zeekr.camerax_android.common.impl

class FocusMeteringActionImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiFocusMeteringActionProxyApi(registrar) {
    class BuilderImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
        PigeonApiFocusMeteringActionBuilderProxyApi(registrar) {
        override fun new1(point: MeteringPoint): FocusMeteringAction.Builder {
            return FocusMeteringAction.Builder(point)
        }

        override fun new2(
            point: MeteringPoint, meteringModes: List<FocusMeteringActionMeteringModeApi>
        ): FocusMeteringAction.Builder {
            return FocusMeteringAction.Builder(point, meteringModes.map { it.impl }.reduce { acc, i -> acc or i })
        }

        override fun addPoint1(
            pigeon_instance: FocusMeteringAction.Builder, point: MeteringPoint
        ): FocusMeteringAction.Builder {
            return pigeon_instance.addPoint(point)
        }

        override fun addPoint2(
            pigeon_instance: FocusMeteringAction.Builder,
            point: MeteringPoint,
            meteringModes: List<FocusMeteringActionMeteringModeApi>
        ): FocusMeteringAction.Builder {
            return pigeon_instance.addPoint(point, meteringModes.map { it.impl }.reduce { acc, i -> acc or i })
        }

        override fun disableAutoCancel(pigeon_instance: FocusMeteringAction.Builder): FocusMeteringAction.Builder {
            return pigeon_instance.disableAutoCancel()
        }

        override fun setAutoCancelDuration(
            pigeon_instance: FocusMeteringAction.Builder, duration: Long, timeUnit: TimeUnitApi
        ): FocusMeteringAction.Builder {
            return pigeon_instance.setAutoCancelDuration(duration, timeUnit.impl)
        }

        override fun build(pigeon_instance: FocusMeteringAction.Builder): FocusMeteringAction {
            return pigeon_instance.build()
        }
    }

    override fun getAutoCancelDurationInMillis(pigeon_instance: FocusMeteringAction): Long {
        return pigeon_instance.autoCancelDurationInMillis
    }

    override fun getMeteringPointsAe(pigeon_instance: FocusMeteringAction): List<MeteringPoint> {
        return pigeon_instance.meteringPointsAe
    }

    override fun getMeteringPointsAf(pigeon_instance: FocusMeteringAction): List<MeteringPoint> {
        return pigeon_instance.meteringPointsAf
    }

    override fun getMeteringPointsAwb(pigeon_instance: FocusMeteringAction): List<MeteringPoint> {
        return pigeon_instance.meteringPointsAwb
    }

    override fun isAutoCancelEnabled(pigeon_instance: FocusMeteringAction): Boolean {
        return pigeon_instance.isAutoCancelEnabled
    }
}

val FocusMeteringActionMeteringModeApi.impl: Int
    get() = when (this) {
        FocusMeteringActionMeteringModeApi.AF -> FocusMeteringAction.FLAG_AF
        FocusMeteringActionMeteringModeApi.AE -> FocusMeteringAction.FLAG_AE
        FocusMeteringActionMeteringModeApi.AWB -> FocusMeteringAction.FLAG_AWB
    }
